cmd/kubeapps-apis/plugins/kapp_controller/packages/v1alpha1/server_data_resources.go

```diff
@@ -399,3 +406,117 @@ func (s *Server) updatePkgInstall(ctx context.Context, cluster, namespace string
 	}
 	return &pkgInstall, nil
 }
+
+// appLabelIdentifier returns the app label identifier for the given Kapp app
+//
+// Apparently, App CRs not being created by a "kapp deploy" command don't have the proper annotations.
+// So, in order to retrieve the annotation value,
+// we have to get the ConfigMap <AppName>-ctrl and, then, fetch the value of the key "labelValue" in data.spec.
+// See https://kubernetes.slack.com/archives/CH8KCCKA5/p1637842398026700
+// https://github.com/vmware-tanzu/carvel-kapp-controller/issues/430
+func (s *Server) appLabelIdentifier(ctx context.Context, cluster, namespace, installedPackageRefId string) (string, error) {
+	typedClient, _, _, err := s.GetClients(ctx, cluster)
+	if err != nil {
+		return "", status.Errorf(codes.Internal, "unable to get the k8s client: '%v'", err)
+	}
+
+	// the ConfigMap name is, by convention, "<appname>-ctrl", but it will change in the near future
+	cmName := fmt.Sprintf("%s-ctrl", installedPackageRefId)
+	cm, err := typedClient.CoreV1().ConfigMaps(namespace).Get(ctx, cmName, metav1.GetOptions{})
+
+	if err != nil || cm.Data["spec"] != "" {
+		log.Warning(errorByStatus("get", "ConfigMap", cmName, err))
+	}
+
+	appLabelValue := extractValueFromJson(cm.Data["spec"], "labelValue")
+
+	return appLabelValue, nil
+}
+
+// findMatchingK8sResources returns the list of k8s resources matching the given listOptions
+// Code inspired by https://github.com/kubernetes/kubectl/blob/release-1.22/pkg/cmd/apiresources/apiresources.go#L142
+func (s *Server) findMatchingK8sResources(ctx context.Context, cluster string, listOptions metav1.ListOptions) ([]*corev1.ResourceRef, error) {
+	_, dynamicClient, discoveryClient, err := s.GetClients(ctx, cluster)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "unable to get the k8s client: '%v'", err)
+	}
+
+	refs := []*corev1.ResourceRef{}
+
+	// fetch every possible kubernetes resource list that is available in the cluster
+	apiResourceLists, err := discoveryClient.ServerPreferredResources()
+	if err != nil {
+		return nil, err
+	}
+
+	// create channel to collect the matching k8s resources thus avoid blocking the waitgroup
+	// ussing a buffered channel with size 100; assumption: 60% native k8s resources + 40 CRDs
+	ch := make(chan resourceAndGvk, 100)
```

Comment:
I'm not used to channels yet, I've made this assumption, but please feel free to tell me more about the buffer sizes here if it is wrong.